File: packages/shared/src/noise.ts
/**
 * Deterministic pixel noise for the noise pen.
 *
 * The value of a pixel depends only on the stroke id and the pixel's *world*
 * coordinates, so a server-side crop (rendered with the crop origin subtracted)
 * produces exactly the same pixels as the client's full-size layer. Nothing
 * here is random: two renders of the same stroke are byte-identical.
 */

/** 32-bit FNV-1a of a string, used to turn a stroke id into a seed. */
export function fnv1a(text: string): number {
  let hash = 0x811c9dc5;
  for (let i = 0; i < text.length; i++) {
    hash ^= text.charCodeAt(i) & 0xff;
    // hash *= 16777619, in 32-bit arithmetic
    hash = Math.imul(hash, 0x01000193) >>> 0;
  }
  return hash >>> 0;
}

/** One 32-bit hash of (seed, x, y); the three low bytes become RGB. */
export function noiseHash(seed: number, x: number, y: number): number {
  let h = (seed ^ Math.imul(x | 0, 0x9e3779b1) ^ Math.imul(y | 0, 0x85ebca77)) >>> 0;
  h = Math.imul(h ^ (h >>> 15), 0x2c1b3c6d) >>> 0;
  h = Math.imul(h ^ (h >>> 12), 0x297a2d39) >>> 0;
  return (h ^ (h >>> 15)) >>> 0;
}

/** Uniform RGB noise for one world pixel. */
export function noiseRGB(seed: number, x: number, y: number): [number, number, number] {
  const h = noiseHash(seed, x, y);
  return [h & 0xff, (h >>> 8) & 0xff, (h >>> 16) & 0xff];
}

File: apps/brushjam/src/brushjam/noise.py
"""Deterministic pixel noise for the noise pen.

Bit-exact port of packages/shared/src/noise.ts: the value of a pixel depends
only on the stroke id and the pixel's *world* coordinates, so a server-side
crop produces exactly the same pixels as the client's full-size layer.
"""

from __future__ import annotations

from typing import Tuple

MASK32 = 0xFFFFFFFF


def _imul(a: int, b: int) -> int:
    """Math.imul: 32-bit signed multiply, returned unsigned."""
    return (a * b) & MASK32


def fnv1a(text: str) -> int:
    """32-bit FNV-1a of a string, over the low byte of each UTF-16 code unit."""
    h = 0x811C9DC5
    for ch in text:
        # charCodeAt(i) & 0xff, i.e. UTF-16 code units, not code points.
        code = ord(ch)
        if code > 0xFFFF:  # astral: JS would see a surrogate pair
            code -= 0x10000
            for unit in (0xD800 + (code >> 10), 0xDC00 + (code & 0x3FF)):
                h ^= unit & 0xFF
                h = _imul(h, 0x01000193)
            continue
        h ^= code & 0xFF
        h = _imul(h, 0x01000193)
    return h & MASK32


def _to_int32(x: int) -> int:
    """JS `x | 0`."""
    x &= MASK32
    return x - 0x1_0000_0000 if x >= 0x8000_0000 else x


def noise_hash(seed: int, x: int, y: int) -> int:
    """One 32-bit hash of (seed, x, y); the three low bytes become RGB."""
    h = (seed ^ _imul(_to_int32(x), 0x9E3779B1) ^ _imul(_to_int32(y), 0x85EBCA77)) & MASK32
    h = _imul(h ^ (h >> 15), 0x2C1B3C6D) & MASK32
    h = _imul(h ^ (h >> 12), 0x297A2D39) & MASK32
    return (h ^ (h >> 15)) & MASK32


def noise_rgb(seed: int, x: int, y: int) -> Tuple[int, int, int]:
    """Uniform RGB noise for one world pixel."""
    h = noise_hash(seed, x, y)
    return (h & 0xFF, (h >> 8) & 0xFF, (h >> 16) & 0xFF)
